src/main/java/com/aws/greengrass/clientdevices/auth/iot/GreengrassClientFactory.java

```diff
@@ -0,0 +1,136 @@
+     * @param lazyCredentialProvider Credential provider for the client
+     */
+    @Inject
+    public GreengrassClientFactory(GreengrassServiceClientFactory greengrassServiceClientFactory,
+                                   LazyCredentialProvider lazyCredentialProvider) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+        this.lazyCredentialProvider = lazyCredentialProvider;
+    }
+
+    /**
+     * Provides a new GG v2 client.
+     * Its upto the caller to maintain the lifecycle of the client.
+     * (TODO: This should ideally exist in Nucleus's GreengrassServiceClientFactory)
+     *
+     * @return GreengrassV2Client
+     */
+    public GreengrassV2Client getGGV2Client()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        GreengrassV2ClientBuilder clientBuilder = GreengrassV2Client.builder()
+                .httpClient(httpClient.build())
+                .credentialsProvider(lazyCredentialProvider)
+                .overrideConfiguration(
+                        ClientOverrideConfiguration.builder().retryPolicy(RetryMode.STANDARD).build());
+
+        if (Utils.isEmpty(awsRegion)) {
+            return clientBuilder.build();
+        }
+        clientBuilder.region(Region.of(awsRegion));
+        getGreengrassServiceEndpoint(deviceConfiguration, awsRegion)
+                .ifPresent(endpoint -> clientBuilder.endpointOverride(URI.create(endpoint)));
+
+        return clientBuilder.build();
+    }
+
+    /**
+     * Provides a new GG v2 Data client without implicit retry policy.
+     * Consumer should handle retries when appropriate and maintain the lifecycle of the client.
+     * The Nucleus provided V2DataClient has a built in retry policy which cannot be overridden.
+     *
+     * @return GreengrassV2DataClient
+     */
+    public GreengrassV2DataClient getGGV2DataClient()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        GreengrassV2DataClientBuilder clientBuilder = GreengrassV2DataClient.builder()
+                .credentialsProvider(lazyCredentialProvider)
+                .httpClient(httpClient.build())
+                .overrideConfiguration(ClientOverrideConfiguration.builder()
+                        .retryPolicy(RetryPolicy.none()).build());
+
+        if (Utils.isEmpty(awsRegion)) {
+            return clientBuilder.build();
+        }
+        clientBuilder.region(Region.of(awsRegion));
+        getGreengrassServiceEndpoint(deviceConfiguration, awsRegion)
```

Comment:
Data and control plane endpoints are not the same, they cannot be using the same API to get the endpoints.

https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/8a8ff7f53eac7a301103969fcbbb2d567e3ccdf8/src/main/java/com/aws/greengrass/componentmanager/ClientConfigurationUtils.java#L49

This is the API that must be used for the data endpoint in order to work properly in VPC.

Reply:
Also need to ensure the right port is being used, since customers may configure it to use 443, rather than the default of 8443.

Reply:
thank you, will fix these.